File: 1298.maximum-candies-you-can-get-from-boxes/solution.go
func maxCandies(status []int, candies []int, keys [][]int, containedBoxes [][]int, initialBoxes []int) int {
    res := 0
    boxes := make([]bool, len(status))
    var queue []int

    for _, b := range initialBoxes {
        if status[b] == 1 {
            queue = append(queue, b)
        } else {
            boxes[b] = true
        }
    }

    for len(queue) > 0 {
        i := queue[0]
        queue = queue[1:]
        res += candies[i]

        for _, b := range containedBoxes[i] {
            if status[b] == 0 {
                boxes[b] = true
            } else { 
                queue = append(queue, b)
            }
        }

        for _, b := range keys[i] {
            if status[b] == 0 {
                if boxes[b] {
                    boxes[b] = false
                    queue = append(queue, b)
                } else {
                    status[b] = 1
                }
            }
        }
    }

    return res
}

File: 1298.maximum-candies-you-can-get-from-boxes/solution.kt
class Solution {
    fun maxCandies(status: IntArray, candies: IntArray, keys: Array<IntArray>, containedBoxes: Array<IntArray>, initialBoxes: IntArray): Int {
        var res = 0
        val boxes = BooleanArray(status.size)
        val queue = ArrayDeque<Int>()

        initialBoxes.forEach { i ->
            if (status[i] == 0) {
                boxes[i] = true
            } else {
                queue.add(i)
            }
        }
        
        while (queue.isNotEmpty()) {
            val i = queue.removeFirst()
            res += candies[i]

            containedBoxes[i].forEach { j ->
                if (status[j] == 0) {
                    boxes[j] = true
                } else {
                    queue.add(j)
                }
            }

            keys[i].forEach { j ->
                if (status[j] == 0) {
                    if (boxes[j]) {
                        boxes[j] = false
                        queue.add(j)
                    } else {
                        status[j] = 1
                    }
                }
            }
        }
        return res
    }
}